src/puppetlabs/puppetdb/middleware.clj

```diff
@@ -298,6 +299,40 @@
           (app req)))
       (app req))))
 
+;; for testing via with-redefs
+(defn get-sync-ver []
+  constants/pdb-sync-ver)
+
+(defn verify-sync-version
+  "Check that the x-pdb-sync-ver header of an incoming requests matches the
+   pdb-sync-ver locally. This header is only sent from pe-puppetdb sync requests
+   and a mismatch indicates that one pdb has upgraded in an incompatible way
+   before the other. In this case all sync requests will return 409s until both
+   sides have been upgraded and have matching sync versions. If the header is
+   present but improperly formatted returns a 400 error response."
+  [app]
+  (fn [{:keys [headers] :as req}]
+    (let [maybe-sync-ver (when-let [v (get headers "x-pdb-sync-ver")]
+                           (try
+                             (Integer/parseInt v)
+                             (catch NumberFormatException e
+                               {:error true :input v})))]
+      (cond
+        (nil? maybe-sync-ver)
```

Comment:
Hmm, wondered about rearranging this a bit, pulling up the common case, checking for the "completely unexpected" case, etc.
```
(cond
  (not maybe-sync-ver) (app req)

  (integer? maybe-sync-ver)
  (if (= maybe-sync-ver (get-sync-ver))
    (app req)
    (error-response...))

  (:error maybe-sync-ver)
  (error-response ...)

  :else something-is-terribly-wrong)
```

Reply:
makes sense, I'll adjust